File: codes/semana12/Streaming_factory.py
from __future__ import annotations
from abc import ABC, abstractmethod
from dataclasses import dataclass
from collections import Counter
import random


# ── Track ────────────────────────────────────────────────
@dataclass(frozen=True)
class Track:
    name:   str
    artist: str
    genre:  str


# ── Strategy: RecommendationEngine ───────────────────────
class RecommendationEngine(ABC):
    @abstractmethod
    def recommend(self, history: list[Track]) -> list[Track]: ...


class SpotifyEngine(RecommendationEngine):
    """Recomenda faixas do gênero mais ouvido."""
    def recommend(self, history: list[Track]) -> list[Track]:
        top_genre = Counter(t.genre for t in history).most_common(1)[0][0]
        return [t for t in history if t.genre == top_genre]


class AppleMusicEngine(RecommendationEngine):
    """Recomenda faixas do artista mais ouvido."""
    def recommend(self, history: list[Track]) -> list[Track]:
        top_artist = Counter(t.artist for t in history).most_common(1)[0][0]
        return [t for t in history if t.artist == top_artist]


class DeezerEngine(RecommendationEngine):
    """Embaralha o histórico e retorna os 3 primeiros."""
    def recommend(self, history: list[Track]) -> list[Track]:
        shuffled = list(history)
        random.shuffle(shuffled)
        return shuffled[:3]


# ── Creator abstrato: MusicPlatform ──────────────────────
class MusicPlatform(ABC):

    def __init__(self) -> None:
        self._cache_key: frozenset | None = None
        self._cache: list[Track] = []

    # Factory Method — subclasse decide qual engine criar
    @abstractmethod
    def create_engine(self) -> RecommendationEngine: ...

    @property
    @abstractmethod
    def platform_name(self) -> str: ...

    def get_playlist(self, history: list[Track]) -> None:
        # Bônus: cache por identidade do histórico
        key = frozenset((t.name, t.artist) for t in history)
        if key == self._cache_key:
            print(f"[cache hit] ", end="")
        else:
            self._cache_key = key
            self._cache = self.create_engine().recommend(history)

        print(f"=== {self.platform_name} — Playlist ===")
        for t in self._cache:
            print(f"  {t.name}  —  {t.artist}")
        print()


# ── Creators concretos ───────────────────────────────────
class SpotifyPlatform(MusicPlatform):
    platform_name = "Spotify"
    def create_engine(self) -> RecommendationEngine:
        return SpotifyEngine()

class AppleMusicPlatform(MusicPlatform):
    platform_name = "Apple Music"
    def create_engine(self) -> RecommendationEngine:
        return AppleMusicEngine()

class DeezerPlatform(MusicPlatform):
    platform_name = "Deezer"
    def create_engine(self) -> RecommendationEngine:
        return DeezerEngine()


# ── Uso ──────────────────────────────────────────────────
if __name__ == "__main__":
    history = [
        Track("Blinding Lights",   "The Weeknd",   "Pop"),
        Track("Starboy",           "The Weeknd",   "Pop"),
        Track("Save Your Tears",   "The Weeknd",   "Pop"),
        Track("Bohemian Rhapsody", "Queen",        "Rock"),
        Track("Back in Black",     "AC/DC",        "Rock"),
        Track("Hotel California",  "Eagles",       "Rock"),
        Track("Shape of You",      "Ed Sheeran",   "Pop"),
        Track("Bad Guy",           "Billie Eilish","Pop"),
    ]

    platforms: list[MusicPlatform] = [
        SpotifyPlatform(),
        AppleMusicPlatform(),
        DeezerPlatform(),
    ]

    for p in platforms:
        p.get_playlist(history)
        # Segunda chamada com mesmo histórico → cache
        p.get_playlist(history)

File: codes/semana12/Streaming_Factory.cpp
#include <iostream>
#include <memory>
#include <string>
#include <vector>
#include <algorithm>
#include <random>
#include <map>
#include <functional>

// ── Track ────────────────────────────────────────────────
struct Track {
    std::string name;
    std::string artist;
    std::string genre;
};

// ── Strategy: RecommendationEngine ───────────────────────
class RecommendationEngine {
public:
    virtual std::vector<Track> recommend(
        const std::vector<Track>& history) const = 0;
    virtual ~RecommendationEngine() = default;
};

// Spotify: recomenda pelo gênero mais ouvido
class SpotifyEngine : public RecommendationEngine {
public:
    std::vector<Track> recommend(
        const std::vector<Track>& history) const override
    {
        // Conta gêneros
        std::map<std::string,int> freq;
        for (const auto& t : history) ++freq[t.genre];
        auto it = std::max_element(freq.begin(), freq.end(),
            [](const auto& a, const auto& b){
                return a.second < b.second; });
        std::string top = it->first;

        std::vector<Track> result;
        for (const auto& t : history)
            if (t.genre == top) result.push_back(t);
        return result;
    }
};

// Apple Music: recomenda pelo artista mais ouvido
class AppleMusicEngine : public RecommendationEngine {
public:
    std::vector<Track> recommend(
        const std::vector<Track>& history) const override
    {
        std::map<std::string,int> freq;
        for (const auto& t : history) ++freq[t.artist];
        auto it = std::max_element(freq.begin(), freq.end(),
            [](const auto& a, const auto& b){
                return a.second < b.second; });
        std::string top = it->first;

        std::vector<Track> result;
        for (const auto& t : history)
            if (t.artist == top) result.push_back(t);
        return result;
    }
};

// Deezer: embaralha e retorna os 3 primeiros
class DeezerEngine : public RecommendationEngine {
public:
    std::vector<Track> recommend(
        const std::vector<Track>& history) const override
    {
        auto shuffled = history;
        std::shuffle(shuffled.begin(), shuffled.end(),
                     std::mt19937{std::random_device{}()});
        if (shuffled.size() > 3) shuffled.resize(3);
        return shuffled;
    }
};

// ── Creator abstrato: MusicPlatform ──────────────────────
class MusicPlatform {
    // Bônus: cache — chave = hash simples do histórico
    mutable std::string          _cacheKey;
    mutable std::vector<Track>   _cache;

    std::string hashHistory(
        const std::vector<Track>& h) const
    {
        std::string key;
        for (const auto& t : h) key += t.name + "|";
        return key;
    }

public:
    // Factory Method
    virtual std::unique_ptr<RecommendationEngine>
        createEngine() const = 0;

    virtual std::string platformName() const = 0;

    void getPlaylist(const std::vector<Track>& history) {
        auto key = hashHistory(history);
        if (key == _cacheKey) {
            std::cout << "[cache hit] ";
        } else {
            _cacheKey = key;
            _cache    = createEngine()->recommend(history);
        }
        std::cout << "=== " << platformName()
                  << " — Playlist ===\n";
        for (const auto& t : _cache)
            std::cout << "  " << t.name
                      << " — " << t.artist << "\n";
        std::cout << "\n";
    }

    virtual ~MusicPlatform() = default;
};

// ── Creators concretos ───────────────────────────────────
class SpotifyPlatform : public MusicPlatform {
public:
    std::unique_ptr<RecommendationEngine>
    createEngine() const override {
        return std::make_unique<SpotifyEngine>();
    }
    std::string platformName() const override {
        return "Spotify"; }
};

class AppleMusicPlatform : public MusicPlatform {
public:
    std::unique_ptr<RecommendationEngine>
    createEngine() const override {
        return std::make_unique<AppleMusicEngine>();
    }
    std::string platformName() const override {
        return "Apple Music"; }
};

class DeezerPlatform : public MusicPlatform {
public:
    std::unique_ptr<RecommendationEngine>
    createEngine() const override {
        return std::make_unique<DeezerEngine>();
    }
    std::string platformName() const override {
        return "Deezer"; }
};

// ── Main ─────────────────────────────────────────────────
int main() {
    std::vector<Track> history = {
        {"Blinding Lights",    "The Weeknd",  "Pop"},
        {"Starboy",            "The Weeknd",  "Pop"},
        {"Save Your Tears",    "The Weeknd",  "Pop"},
        {"Bohemian Rhapsody",  "Queen",       "Rock"},
        {"Back in Black",      "AC/DC",       "Rock"},
        {"Hotel California",   "Eagles",      "Rock"},
        {"Shape of You",       "Ed Sheeran",  "Pop"},
        {"Bad Guy",            "Billie Eilish","Pop"},
    };

    std::vector<std::unique_ptr<MusicPlatform>> platforms;
    platforms.push_back(std::make_unique<SpotifyPlatform>());
    platforms.push_back(std::make_unique<AppleMusicPlatform>());
    platforms.push_back(std::make_unique<DeezerPlatform>());

    for (auto& p : platforms) {
        p->getPlaylist(history);
        // Segunda chamada → deve usar cache (Spotify/Apple)
        p->getPlaylist(history);
    }
    return 0;
}